Fix #4246: RTL text plugin caused leak on Style#_remove()

ChrisLoer opened the pull request:
The unit test verifies that the "deregister" call happens, but because actually exercising the plugin-loading code depends on the network, I fell back to manual testing.

I tested by creating two maps, verifying there were two callbacks registered, removing one map and seeing that the second callback was removed, then calling setRTLTextPlugin to verify the remaining callback was successfully called and then discarded.

@jfirebaugh @mourner

lucaswoj commented:
This is starting to feel like an implementation of [`Evented`](https://www.mapbox.com/mapbox-gl-js/api/#Evented). Would it be possible to use `Evented` to manage plugin available callbacks? That code is well tested and believed to be memory-leak free.

ChrisLoer commented:
@lucaswoj I switched to using`Evented`, but had to modify the implementation `Evented.once()` so that `off()` works on one-time listeners as well as listeners added with `on()`.

I also modified `Evented.listens()` so that it wouldn't return true when listeners had been added and then removed. The rest of this PR doesn't depend on that change, but at least based on my reading of the function documentation, it's the expected behavior.
